cmd/docker-proxy/udp_proxy_linux.go

```diff
@@ -51,26 +54,37 @@ type UDPProxy struct {
 	backendAddr    *net.UDPAddr
 	connTrackTable connTrackMap
 	connTrackLock  sync.Mutex
+	ipVer          ipVersion
 }
 
 // NewUDPProxy creates a new UDPProxy.
-func NewUDPProxy(listener *net.UDPConn, backendAddr *net.UDPAddr) (*UDPProxy, error) {
+func NewUDPProxy(listener *net.UDPConn, backendAddr *net.UDPAddr, ipVer ipVersion) (*UDPProxy, error) {
 	return &UDPProxy{
 		listener:       listener,
 		frontendAddr:   listener.LocalAddr().(*net.UDPAddr),
 		backendAddr:    backendAddr,
 		connTrackTable: make(connTrackMap),
+		ipVer:          ipVer,
 	}, nil
 }
 
-func (proxy *UDPProxy) replyLoop(proxyConn *net.UDPConn, clientAddr *net.UDPAddr, clientKey *connTrackKey) {
+func (proxy *UDPProxy) replyLoop(proxyConn *net.UDPConn, serverAddr net.IP, clientAddr *net.UDPAddr, clientKey *connTrackKey) {
 	defer func() {
 		proxy.connTrackLock.Lock()
 		delete(proxy.connTrackTable, *clientKey)
 		proxy.connTrackLock.Unlock()
 		proxyConn.Close()
 	}()
 
+	var oob []byte
+	if proxy.ipVer == ip4 {
+		cm := &ipv4.ControlMessage{Src: serverAddr}
+		oob = cm.Marshal()
+	} else {
+		cm := &ipv6.ControlMessage{Src: serverAddr}
+		oob = cm.Marshal()
```

Comment:
`ipv4.NewControlMessage()` / `ipv6.NewControlMessage()` both takes a single argument specifying which types of ancillary data we want and returns a `[]byte` buffer of the appropriate size. This is used to read control messages.

Right-sizing the buffer passed to `recvmsg` is somewhat important because the sysctl `net.core.optmem_max` defines the maximum amount of ancillary data each socket can hold.

So, nope, we can't use `NewControlMessage` to return a buffer containing options.